tests failing on master (macOS Monterey / py 3.7)

**Bug Report**

Tests are failing for the master branch on my macOS laptop.

A few tests in `PEP561Suite` are flaky for reasons I don't quite understand. 

**To Reproduce**

```
git pull
git checkout f501cf649d7976077a7196e3548d773d67340a8c
python3.7 -m venv .venv
.venv/bin/pip install -r test-requirements.txt
.venv/bin/pip install -e .
PATH=$(pwd)/.venv/bin:$PATH .venv/bin/pytest -q -k 'PEP561Suite'
```

**Expected Behavior**

Tests pass.

**Actual Behavior**

There's usually exactly one failure per run. On some rare occasions I've seen more than one failure, or no failures.

Here are some sample failures:

```
_____________________________________________________________________________________________________________ testStubPrecedence _____________________________________________________________________________________________________________
[gw3] darwin -- Python 3.7.13 /System/Volumes/Data/repos/mypy/.venv/bin/python3.7
data: /System/Volumes/Data/repos/mypy/test-data/unit/pep561.test:66:
/System/Volumes/Data/repos/mypy/mypy/test/testpep561.py:30: in run_case
    test_pep561(test_case)
/System/Volumes/Data/repos/mypy/mypy/test/testpep561.py:152: in test_pep561
    testcase.file, testcase.line, iter_count))
E   AssertionError: Invalid output (/System/Volumes/Data/repos/mypy/test-data/unit/pep561.test, line 66)
------------------------------------------------------------------------------------------------------------ Captured stderr call ------------------------------------------------------------------------------------------------------------
Expected:
  testStubPrecedence.py:5: note: Revealed type is "builtins.list[builtins.str]" (diff)
Actual:
  testStubPrecedence.py:3: error: Module "typedpkg" has no attribute "dne" (diff)
  testStubPrecedence.py:5: note: Revealed type is "builtins.list[builtins.str]" (diff)

Alignment of first line difference:
  E: testStubPrecedence.py:5: note: Revealed type is "builtins.list[builtins....
  A: testStubPrecedence.py:3: error: Module "typedpkg" has no attribute "dne"...
                           ^
```

```
______________________________________________________________________________________________ testNamespacePkgWStubsWithNamespacePackagesFlag _______________________________________________________________________________________________
[gw2] darwin -- Python 3.7.13 /System/Volumes/Data/repos/mypy/.venv/bin/python3.7
data: /System/Volumes/Data/repos/mypy/test-data/unit/pep561.test:204:
/System/Volumes/Data/repos/mypy/mypy/test/testpep561.py:30: in run_case
    test_pep561(test_case)
/System/Volumes/Data/repos/mypy/mypy/test/testpep561.py:152: in test_pep561
    testcase.file, testcase.line, iter_count))
E   AssertionError: Invalid output (/System/Volumes/Data/repos/mypy/test-data/unit/pep561.test, line 204)
------------------------------------------------------------------------------------------------------------ Captured stderr call ------------------------------------------------------------------------------------------------------------
Expected:
  testNamespacePkgWStubsWithNamespacePackagesFlag.py:7: error: Argument 1 to "bf" has incompatible type "int"; expected "bool" (diff)
  testNamespacePkgWStubsWithNamespacePackagesFlag.py:8: error: Argument 1 to "bf" has incompatible type "int"; expected "bool" (diff)
Actual:
  testNamespacePkgWStubsWithNamespacePackagesFlag.py:3: error: Skipping analyzing "typedpkg_ns.a.bbb": module is installed, but missing library stubs or py.typed marker (diff)
  testNamespacePkgWStubsWithNamespacePackagesFlag.py:3: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports (diff)
  testNamespacePkgWStubsWithNamespacePackagesFlag.py:3: error: Skipping analyzing "typedpkg_ns.a": module is installed, but missing library stubs or py.typed marker (diff)
  testNamespacePkgWStubsWithNamespacePackagesFlag.py:8: error: Argument 1 to "bf" has incompatible type "int"; expected "bool" (diff)

Alignment of first line difference:
  E: ...spacePackagesFlag.py:7: error: Argument 1 to "bf" has incompatible ty...
  A: ...spacePackagesFlag.py:3: error: Skipping analyzing "typedpkg_ns.a.bbb"...
                             ^
```

```
_______________________________________________________________________________________________________ testTypedPkgNamespaceRegImport _______________________________________________________________________________________________________
[gw2] darwin -- Python 3.7.13 /System/Volumes/Data/repos/mypy/.venv/bin/python3.7
data: /System/Volumes/Data/repos/mypy/test-data/unit/pep561.test:150:
/System/Volumes/Data/repos/mypy/mypy/test/testpep561.py:30: in run_case
    test_pep561(test_case)
/System/Volumes/Data/repos/mypy/mypy/test/testpep561.py:152: in test_pep561
    testcase.file, testcase.line, iter_count))
E   AssertionError: Invalid output (/System/Volumes/Data/repos/mypy/test-data/unit/pep561.test, line 150)
------------------------------------------------------------------------------------------------------------ Captured stderr call ------------------------------------------------------------------------------------------------------------
Expected:
  testTypedPkgNamespaceRegImport.py:4: error: Cannot find implementation or library stub for module named "typedpkg_ns.a.dne" (diff)
  testTypedPkgNamespaceRegImport.py:4: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports (diff)
  testTypedPkgNamespaceRegImport.py:10: error: Argument 1 has incompatible type "bool"; expected "str" (diff)
  testTypedPkgNamespaceRegImport.py:11: error: Argument 1 has incompatible type "int"; expected "bool" (diff)
Actual:
  testTypedPkgNamespaceRegImport.py:3: error: Cannot find implementation or library stub for module named "typedpkg_ns.a.bbb" (diff)
  testTypedPkgNamespaceRegImport.py:3: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports (diff)
  testTypedPkgNamespaceRegImport.py:3: error: Cannot find implementation or library stub for module named "typedpkg_ns" (diff)
  testTypedPkgNamespaceRegImport.py:3: error: Cannot find implementation or library stub for module named "typedpkg_ns.a" (diff)
  testTypedPkgNamespaceRegImport.py:4: error: Cannot find implementation or library stub for module named "typedpkg_ns.a.dne" (diff)
  testTypedPkgNamespaceRegImport.py:10: error: Argument 1 has incompatible type "bool"; expected "str" (diff)

Alignment of first line difference:
  E: ...kgNamespaceRegImport.py:4: error: Cannot find implementation or libra...
  A: ...kgNamespaceRegImport.py:3: error: Cannot find implementation or libra...
                                ^
```

```
_______________________________________________________________________________________________________ testTypedPkgNamespaceRegImport _______________________________________________________________________________________________________
[gw0] darwin -- Python 3.7.13 /System/Volumes/Data/repos/mypy/.venv/bin/python3.7
data: /System/Volumes/Data/repos/mypy/test-data/unit/pep561.test:150:
/System/Volumes/Data/repos/mypy/mypy/test/testpep561.py:30: in run_case
    test_pep561(test_case)
/System/Volumes/Data/repos/mypy/mypy/test/testpep561.py:113: in test_pep561
    install_package(pkg, python_executable, use_pip, editable)
/System/Volumes/Data/repos/mypy/mypy/test/testpep561.py:86: in install_package
    raise Exception(proc.stdout.decode('utf-8') + proc.stderr.decode('utf-8'))
E   Exception: Looking in indexes: https://registry.affirm-stage.com/artifactory/api/pypi/pypi/simple
E   Processing /System/Volumes/Data/repos/mypy/test-data/packages/typedpkg_ns_a
E     Preparing metadata (setup.py): started
E     Preparing metadata (setup.py): finished with status 'done'
E   Building wheels for collected packages: typedpkg-namespace.alpha
E     Building wheel for typedpkg-namespace.alpha (setup.py): started
E     Building wheel for typedpkg-namespace.alpha (setup.py): finished with status 'done'
E     Created wheel for typedpkg-namespace.alpha: filename=typedpkg_namespace.alpha-1.0.0-py3-none-any.whl size=1588 sha256=21e162c192f4c7f108ec1b970a54d677de03e021e7143da28589e81f308a7505
E     Stored in directory: /private/var/folders/kd/x5k73mjs4jn02v7k1c0spz_40000gn/T/pip-ephem-wheel-cache-niwbfkch/wheels/85/1b/b9/19a1923b949574d3f92d0b2fd98b06cfe4eb7aa0b819bbbcf7
E   ERROR: Exception:
E   Traceback (most recent call last):
E     File "/var/folders/kd/x5k73mjs4jn02v7k1c0spz_40000gn/T/tmptvamo6c8/lib/python3.7/site-packages/pip/_vendor/pkg_resources/__init__.py", line 2681, in version
E       return self._version
E     File "/var/folders/kd/x5k73mjs4jn02v7k1c0spz_40000gn/T/tmptvamo6c8/lib/python3.7/site-packages/pip/_vendor/pkg_resources/__init__.py", line 2815, in __getattr__
E       raise AttributeError(attr)
E   AttributeError: _version
E
E   During handling of the above exception, another exception occurred:
E
E   Traceback (most recent call last):
E     File "/var/folders/kd/x5k73mjs4jn02v7k1c0spz_40000gn/T/tmptvamo6c8/lib/python3.7/site-packages/pip/_internal/cli/base_command.py", line 167, in exc_logging_wrapper
E       status = run_func(*args)
E     File "/var/folders/kd/x5k73mjs4jn02v7k1c0spz_40000gn/T/tmptvamo6c8/lib/python3.7/site-packages/pip/_internal/cli/req_command.py", line 205, in wrapper
E       return func(self, options, args)
E     File "/var/folders/kd/x5k73mjs4jn02v7k1c0spz_40000gn/T/tmptvamo6c8/lib/python3.7/site-packages/pip/_internal/commands/install.py", line 366, in run
E       global_options=[],
E     File "/var/folders/kd/x5k73mjs4jn02v7k1c0spz_40000gn/T/tmptvamo6c8/lib/python3.7/site-packages/pip/_internal/wheel_builder.py", line 354, in build
E       req.editable and req.permit_editable_wheels,
E     File "/var/folders/kd/x5k73mjs4jn02v7k1c0spz_40000gn/T/tmptvamo6c8/lib/python3.7/site-packages/pip/_internal/wheel_builder.py", line 227, in _build_one
E       _verify_one(req, wheel_path)
E     File "/var/folders/kd/x5k73mjs4jn02v7k1c0spz_40000gn/T/tmptvamo6c8/lib/python3.7/site-packages/pip/_internal/wheel_builder.py", line 175, in _verify_one
E       dist_verstr = str(dist.version)
E     File "/var/folders/kd/x5k73mjs4jn02v7k1c0spz_40000gn/T/tmptvamo6c8/lib/python3.7/site-packages/pip/_internal/metadata/pkg_resources.py", line 147, in version
E       return parse_version(self._dist.version)
E     File "/var/folders/kd/x5k73mjs4jn02v7k1c0spz_40000gn/T/tmptvamo6c8/lib/python3.7/site-packages/pip/_vendor/pkg_resources/__init__.py", line 2689, in version
E       raise ValueError(msg, self)
E   ValueError: ("Missing 'Version:' header and/or METADATA file at path: [could not detect]", typedpkg-namespace-alpha [unknown version] (/private/var/folders/kd/x5k73mjs4jn02v7k1c0spz_40000gn/T/pip-ephem-wheel-cache-niwbfkch/wheels/85/1b/b9/19a1923b949574d3f92d0b2fd98b06cfe4eb7aa0b819bbbcf7/typedpkg_namespace.alpha-1.0.0-py3-none-any.whl))
```

**Your Environment**

mypy commit `f501cf649d7976077a7196e3548d773d67340a8c`
Python 3.7.13
macOS Monterey 12.3.1
x86_64

--- a/.gitignore
+++ b/.gitignore
@@ -14,6 +14,7 @@ venv/
 .mypy_cache/
 .incremental_checker_cache.json
 .cache
+test-data/packages/.pip_lock
 dmypy.json
 .dmypy.json
 

--- a/mypy/test/config.py
+++ b/mypy/test/config.py
@@ -15,3 +15,12 @@ package_path = os.path.join(PREFIX, 'test-data', 'packages')
 # This is *within* the tempfile.TemporaryDirectory that is chroot'ed per testcase.
 # It is also hard-coded in numerous places, so don't change it.
 test_temp_dir = 'tmp'
+
+# The PEP 561 tests do a bunch of pip installs which, even though they operate
+# on distinct temporary virtual environments, run into race conditions on shared
+# file-system state. To make this work reliably in parallel mode, we'll use a
+# FileLock courtesy of the tox-dev/py-filelock package.
+# Ref. https://github.com/python/mypy/issues/12615
+# Ref. mypy/test/testpep561.py
+pip_lock = os.path.join(package_path, '.pip_lock')
+pip_timeout = 60

--- a/mypy/test/testpep561.py
+++ b/mypy/test/testpep561.py
@@ -1,4 +1,5 @@
 from contextlib import contextmanager
+import filelock
 import os
 import pytest
 import re
@@ -9,7 +10,7 @@ import tempfile
 from typing import Tuple, List, Generator
 
 import mypy.api
-from mypy.test.config import package_path
+from mypy.test.config import package_path, pip_lock, pip_timeout
 from mypy.util import try_find_python2_interpreter
 from mypy.test.data import DataDrivenTestCase, DataSuite
 from mypy.test.config import test_temp_dir
@@ -77,11 +78,15 @@ def install_package(pkg: str,
         env = {'PIP_BUILD': dir}
         # Inherit environment for Windows
         env.update(os.environ)
-        proc = subprocess.run(install_cmd,
-                              cwd=working_dir,
-                              stdout=PIPE,
-                              stderr=PIPE,
-                              env=env)
+        try:
+            with filelock.FileLock(pip_lock, timeout=pip_timeout):
+                proc = subprocess.run(install_cmd,
+                                      cwd=working_dir,
+                                      stdout=PIPE,
+                                      stderr=PIPE,
+                                      env=env)
+        except filelock.Timeout as err:
+            raise Exception("Failed to acquire {}".format(pip_lock)) from err
     if proc.returncode != 0:
         raise Exception(proc.stdout.decode('utf-8') + proc.stderr.decode('utf-8'))
 

--- a/runtests.py
+++ b/runtests.py
@@ -58,11 +58,15 @@ cmds = {
     'pytest-slow': ['pytest', '-q', '-k', ' or '.join(
         [SAMPLES,
          TYPESHED,
-         PEP561,
          DAEMON,
          MYPYC_EXTERNAL,
          MYPYC_COMMAND_LINE,
          ERROR_STREAM])],
+
+    # Test cases that might take minutes to run
+    'pytest-slower': ['pytest', '-q', '-k', ' or '.join(
+        [PEP561])],
+
     # Test cases to run in typeshed CI
     'typeshed-ci': ['pytest', '-q', '-k', ' or '.join([CMDLINE,
                                                        EVALUATION,

--- a/test-requirements.txt
+++ b/test-requirements.txt
@@ -1,6 +1,8 @@
 -r mypy-requirements.txt
 -r build-requirements.txt
 attrs>=18.0
+filelock>=3.0.0,<3.4.2; python_version<'3.7'
+filelock>=3.0.0; python_version>='3.7'
 flake8==3.9.2
 flake8-bugbear==22.3.20
 flake8-pyi>=20.5

